### PR TITLE
CMake build system.

### PR DESCRIPTION
**Pull request by coldtobi**
Hi @hzeller,

I could not resist and have a peek at the project, and in preparation for Debian inclusion I've hacked together CMake integration.

My selfish motivations: While Makefiles for works nice, my experience shows that with the many architectures Debian supports this tends to be more fragile. 

I've tries hard to have CMake nehave exactly as src/Makefile… However, here are differences:
- when the version information can be retrieved from git, the timestamp is according
  ISO8601, not only the date. As side effect, one can see if it as cmake or makefile build:))
- The version generation can be turned off with a CMake option -- useful when a distribution does
  package from a git repo unrelated to the upstream git repo.)
- if the manpage is regenerated, the manpage is not compressed.
)

The change in the Makefile makes sure that the CMake and Makefile can coexists; if you think Makefile suppor is not needed, I can simply stuff a bit. Just let me know.

Cheers,
tobi

### CODE REVIEW

**Review by hzeller — CHANGES_REQUESTED** [review]
Nice!

I think this looks like a good move. I am also open to remove the Makefiles entirely; I just got it started as it was simple. Cmake is way more complicated (and I have yet to start trying to understand it), but I agree that it is nicer to get things going more platform independently. So feel free to entirely remove the Makefiles in this pull request, which will make things a little bit simpler (like the workaround the version).

The github cmake stuff looks like a blob of stuff, but given that it is hidden away in the cmake directory it is probably ok.

I don't mind the more 'complete' timestamp.

Back in the day, we always gzipped manpages, but looks like these days that is not common anymore. So I am happy to not have them gzipped (and also open to check in timg.1 instead of timg.1.gz)

One request I'd have: In #37 some documentation was added how to compile it on a Mac using the makefile (some LDFLAGS was needed). If you have an access to a Mac, can you find out if the cmake compilation will make it work on a brew-based system without extra things to do ?

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,27 @@
+cmake_minimum_required(VERSION 3.13)
+
+project(timg VERSION 0.9.9 LANGUAGES CXX)
+
+option(WITH_VIDEO_DECODING "Enables video decoding feature" OFF)
```
[the remark this answers is not in the document]
Comment (coldtobi) [reply CMakeLists.txt]: :+1: changed.

**Comment by coldtobi** [comment]
> Nice!
> 
> I think this looks like a good move. I am also open to remove the Makefiles entirely; I just got it started as it was simple. Cmake is way more complicated (and I have yet to start trying to understand it), but I agree that it is nicer to get things going more platform independently. So feel free to entirely remove the Makefiles in this pull request, which will make things a little bit simpler (like the workaround the version).

Yes, CMake is a beast of its kind… It happens years after using it you still scratch your head sometimes… But IMHO it is currently the option with least head scratching… 

> 
> The github cmake stuff looks like a blob of stuff, but given that it is hidden away in the cmake directory it is probably ok.

Yeah, the good thing about cmake are that there are many modules… I'm using this module at work (since a few years) as well and had no problems yet. Upstream of that module is still active, so I hope this module will create no extra work…

> I don't mind the more 'complete' timestamp.

✔
 
> Back in the day, we always gzipped manpages, but looks like these days that is not common anymore. So I am happy to not have them gzipped 
✔
 
> (and also open to check in timg.1 instead of timg.1.gz)

Another possiblity, as some people say, generated stuff should not be in a git repo:
Only generate the manpage if ronn is installed. Maybe three options:
- Have a options to disable man page generation with defaulting to ON
- Expect users to have ronn installed and fail the build if not
- If ronn is not there, skip manpage generation/installatio.

> One request I'd have: In #37 some documentation was added how to compile it on a Mac using the makefile (some LDFLAGS was needed). If you have an access to a Mac, can you find out if the cmake compilation will make it work on a brew-based system without extra things to do ?

Unfortunately I do not have access to a Mac as well… But maybe the poster of #37 has and we could ask them to check?
(I think there is a good change that cmake will work out of the box…)

**Comment by speedy-beaver** [comment]
[..]
> > One request I'd have: In #37 some documentation was added how to compile it on a Mac using the makefile (some LDFLAGS was needed). If you have an access to a Mac, can you find out if the cmake compilation will make it work on a brew-based system without extra things to do ?
> 
> Unfortunately I do not have access to a Mac as well… But maybe the poster of #37 has and we could ask them to check?
> (I think there is a good change that cmake will work out of the box…)

I posted #37 - can try to verify cmake on Mac. I have cloned the coldtobi:cmake branch - how is cmake building supposed to work? Can you give me some guidance, respectively just spell out what you do on Linux to get it to build?

**Comment by coldtobi** [comment]
Am 7. Februar 2021 13:22:46 MEZ schrieb speedy-beaver <notifications@github.com>:
>[..]>
>> > One request I'd have: In #37 some documentation was added how to
>compile it on a Mac using the makefile (some LDFLAGS was needed). If
>you have an access to a Mac, can you find out if the cmake compilation
>will make it work on a brew-based system without extra things to do ?>
>> >
>> Unfortunately I do not have access to a Mac as well… But maybe the
>poster of #37 has and we could ask them to check?>
>> (I think there is a good change that cmake will work out of the
>box…)>
>>
>I posted #37 - can try to verify cmake on Mac. I have cloned the
>coldtobi:cmake branch - how is cmake building supposed to work? Can you
>give me some guidance, respectively just spell out what you do on Linux
>to get it to build?>
>>
>>
>-- >
>You are receiving this because you authored the thread.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/hzeller/timg/pull/38#issuecomment-774665760

thanks @speedy-beaver 

on linux i'd do:
(in the git checked out dir)

mkdir build
cd build
cmake ..
make

optionally you can try install with

make install DESTDIR=tmp

(DESTDIR will install to that directory, to avoid messing up your system.)

tobi

**Comment by speedy-beaver** [comment]
@coldtobi Thank you, instructions worked, managed to build the timg binary. 

The man page is not getting build because ronn is not available. Homebrew don't seem to have it as part of core and the user contributed tap to install ronn-ng failed to deploy on my machine. Are there alternatives to ronn or is this Mac being lightyears behind?

**Comment by hzeller** [comment]
So given that `ronn` is not available on the Mac brew install (yet?) and that possibly not many users - even the ones that are `build-essentials`-savvy have that installed on their system (I am thinking of the users who just download the sources tar.gz and quickly build it), maybe we should ship the pre-built manpage for now.

Also, I am not tied to `ronn` - it was just the first thing that I came accross as I didn't want to write `roff(7)` and like Markdown as it can be rendered by github. Maybe you know a better/more common way to convert Markdown to manpages @coldtobi ?

**Comment by coldtobi** [comment]
> So given that `ronn` is not available on the Mac brew install (yet?) and that possibly not many users - even the ones that are `build-essentials`-savvy have that installed on their system (I am thinking of the users who just download the sources tar.gz and quickly build it), maybe we should ship the pre-built manpage for now.

Yes, I think this is sane. (Didn't expect ronn to be a bigger problem, but well… )

> Also, I am not tied to `ronn` - it was just the first thing that I came accross as I didn't want to write `roff(7)` and like Markdown as it can be rendered by github. Maybe you know a better/more common way to convert Markdown to manpages @coldtobi ?

Not used by myself, but I heard people use pandoc quote often these days.

**Comment by hzeller** [comment]
@speedy-beaver , is pandoc available on the Mac ?

**Comment by speedy-beaver** [comment]
pandoc is available for Mac from various sources - either directly from pandoc.org as a package or through homebrew. Running the packaged installation from pandoc.org on Apple Silicon now, manages to create a man page, the current timg.1.md will need some reformatting though.

**Comment by dankamongmen** [comment]
having tried both ronn and pandoc this past year for Notcurses, i ended up strongly preferring pandoc, though it is not without its own problems -- i find that it's uninstallable on various distributions from time to time (debian is pretty good, fedora less so, freebsd less than that). also, the haskell toolchain is pretty large if that's something that worries you. nonetheless, i continue to use pandoc (i've got about thirty man pages [here](https://notcurses.com)

**Comment by coldtobi** [comment]
obsolete now with #40 merged.

**Pull request closed without merging**